lib/trie/trie.go

```diff
@@ -54,28 +60,42 @@ func (t *Trie) Snapshot() (newTrie *Trie) {
 	rootCopySettings.CopyCached = true
 	for rootHash, childTrie := range t.childTries {
 		childTries[rootHash] = &Trie{
-			generation:          childTrie.generation + 1,
-			root:                childTrie.root.Copy(rootCopySettings),
-			deletedMerkleValues: make(map[string]struct{}),
+			generation:             childTrie.generation + 1,
+			root:                   childTrie.root.Copy(rootCopySettings),
+			deletedMerkleValues:    make(map[string]struct{}),
+			newDeletedMerkleValues: make(map[string]struct{}),
 		}
 	}
 
 	return &Trie{
-		generation:          t.generation + 1,
-		root:                t.root,
-		childTries:          childTries,
-		deletedMerkleValues: make(map[string]struct{}),
+		generation:             t.generation + 1,
+		root:                   t.root,
+		childTries:             childTries,
+		deletedMerkleValues:    make(map[string]struct{}),
+		newDeletedMerkleValues: make(map[string]struct{}),
 	}
 }
 
+// handleTrackedDeltas modifies the trie deleted merkle values set
+// only if the error is nil. In all cases, it resets the
+// `newDeletedMerkleValues` map.
+func (t *Trie) handleTrackedDeltas(err error) {
+	if err == nil {
+		for merkleValue := range t.newDeletedMerkleValues {
+			t.deletedMerkleValues[merkleValue] = struct{}{}
+		}
+	}
```

Comment:
this seems really awkward.  Could you not just call this whenever `err == nil`, and explicitly reset the `t.newDeletedMerkleValues`.

Reply:
Well I get your point, although this function is about to be stuffed with more logic to track inserted merkle values and remove re-inserted merkle values too. Maybe even more stuff with lazy loading.